File: iteration2/Models/Advisor.java
package iteration2.Models;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class Advisor extends Person {
    private ArrayList<Student> students;
    private RegistrationError error;

    private static String file_path="/iteration2/Logs/Advisor.log";
    private static Logger logger=Logger.getLogger(Advisor.class.getName());
    private static FileHandler fileHandler;

    static {
        try {
            fileHandler = new FileHandler(System.getProperty("user.dir")+file_path,true);
            logger.addHandler(fileHandler);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void customLog(boolean type,String message){
        SimpleFormatter formatter =new SimpleFormatter();
        fileHandler.setFormatter(formatter);
        if(type)
            logger.info(message);
        else
            logger.warning(message);
    }

    public Advisor(ArrayList<Student> students, String name, String surname, String ssn, Character gender, RegistrationError error) {
        super(name,surname,ssn,gender);
        this.students = students;
        this.error = error;
        customLog(true, "Advisor constructor method is called and new advisor object is generated.");
    }

    public boolean courseAvailability(Student student, Course course) {

        if (!checkSemester(student, course)) {
            student.addError(this.error.reportError(1001,new String[]{course.getName(),Integer.toString(course.getSemester()), Integer.toString(student.getSemesterNo())}));
            customLog(false, "Checked semester of students and courses that requested by students and they are declined ");
            return false;
        }
        if(!checkCredit(student, course)) {
            student.addError(this.error.reportError(1002,new String[]{course.getName(),Integer.toString(student.getTranscript().getCompletedCredit())}));
            customLog(false, "Compared credit of students and required credit of course and it is not enough.");
            return false;
        }
        if(!checkQuota(course)) {
            student.addError(this.error.reportError(1003,new String[]{course.getName()}));
            customLog(false, "Check quota of course and there is not enough space for student.");
            return false;
        }
        Object[] tempReturn = checkPreRequisite(student, course);
        if(!(boolean)tempReturn[0]) {
            student.addError(this.error.reportError(1004,new String[]{course.getName(),(String)tempReturn[1]}));
            //return false;
        }
        Object[] tempReturn2 = checkCollision(student);
        if(!(boolean)tempReturn2[0]) {
            student.addError(this.error.reportError(1005,new String[]{(String)tempReturn2[1],(String)tempReturn2[2],
                    (String)tempReturn2[3]}));
            //return false;
        }
        if(!checkElective(student, course)) {
            //1005
        }
        if(!FTETakeable(student, course)) {
            //1006
        }
        customLog(true, "courseAvailability method is called and advisor check the course availability.");
        return true;
    }

    public boolean checkSemester(Student student, Course course) {
        if(course.getSemester() <= student.getSemesterNo()) {
            return true;
        }
        return false;
    }

    public boolean checkQuota(Course course) {
        if(course.getRegisteredStudentsCount() >= course.getQuota()) {
            return false;
        }
        return true;
    }
    public Object[] checkPreRequisite(Student student, Course course) {
        ArrayList<Course> tempCompletedCourse = student.getTranscript().getCompletedCourses();
        boolean availability = true;
        Object[] errorInfo = new Object[2];
        for (int i=0; i<course.getPreRequisiteCourses().size();i++){
            for (int j = 0; j<tempCompletedCourse.size();j++){
                //TO-DO : check if the course is in the completed courses
                if (course.getPreRequisiteCourses().get(i).getCode()!=tempCompletedCourse.get(j).getCode()){
                    availability = false;
                    errorInfo[1] = course.getPreRequisiteCourses().get(i).getCode();
                }
                else {
                    availability = true;
                    errorInfo[1] = "";
                    break;
                }
            }
            if(!availability){break;}
        }
        errorInfo[0] = availability;
        return errorInfo;
    }

    public boolean checkCredit(Student student, Course course) {
        if(course.getRequiredCredits() > student.getTranscript().getCompletedCredit()) {
            return false;
        }
        return true;
    }

    public Object[] checkCollision(Student student) {
        Set<Course> coursesOfHash = student.getSelectedCourses().keySet();
        ArrayList<Course> courses = new ArrayList<>(coursesOfHash);
        Object[] errorInfo = new Object[4];
        errorInfo[0] = true;
        errorInfo[1] = "";
        errorInfo[2] = "";
        errorInfo[3] = "";
        for(int i=0; i<courses.size(); i++){
            ArrayList<String> temp = new ArrayList<>(courses.get(i).getWeeklyHours());
            for (int j = i+1; j< courses.size();j++){
                temp.retainAll(courses.get(j).getWeeklyHours());
                if (temp.size() != 0){
                    errorInfo[0] = false;
                    errorInfo[1] = courses.get(i).getName();
                    errorInfo[2] = courses.get(j).getName();
                    errorInfo[3] = Integer.toString(temp.size());
                    return errorInfo;
                }
            }
        }
        return errorInfo;
    }

    public boolean checkElective(Student student, Course course) {
        AtomicInteger count = new AtomicInteger();
        student.getSelectedCourses().forEach((k,v)->{
            if (k.getClass().equals(Elective.class))
                count.getAndIncrement();


        });
        if (count.get() >=2){
            return false;
        }
        return true;
    }

    public boolean FTETakeable(Student student, Course course) {
        if(student.getSemesterNo()>= 8) {
            return true;
        }
        return false;
    }

    public ArrayList<Student> getStudents() {
        return this.students;
    }

    public void setStudents(ArrayList<Student> students) {
        for (int i=0; i<students.size();i++){
            for (int j = 0; j<this.students.size();j++){
                if (students.get(i).getId()!=this.students.get(j).getId()){
                    this.students.add(students.get(i));
                }
            }
        }
    }

    public JSONArray convertArraylistToJsonArray(ArrayList<Student> arrayList){
        JSONArray jsonArray=new JSONArray();
        for (int i=0;i<arrayList.size();i++){
            jsonArray.put(arrayList.get(i).getName() + " " + arrayList.get(i).getSurname());
        }
        return jsonArray;
    }

    @Override
    public JSONObject toJson() {
        JSONObject jsonObject=new JSONObject();
        jsonObject.put("students",convertArraylistToJsonArray(getStudents()));
        jsonObject.put("name",getName());
        jsonObject.put("surname",getSurname());
        jsonObject.put("ssn",getSsn());
        jsonObject.put("gender",getGender());
        return jsonObject;
    }
}

File: iteration3/Models/Advisor.py
from abc import ABC

from iteration3.Models import Elective, ElectiveType
from iteration3.Models.Course import Course
from iteration3.Models.Person import Person
from iteration3.Models.Student import Student


class Advisor(Person):
    def __init__(self, name, surname, ssn, gender, students, error):
        super().__init__(name, surname, ssn, gender)
        self.__students = students
        self.__error = error

        def to_json(self):
            return ""

    def custom_log(self, is_error, message):
        if is_error:
            self.logger.info(message)
        else:
            self.logger.warning(message)

    def course_availability(self, student:Student, course:Course):
        if not self.check_semester(self, student, course):
            student.add_error(
                self._error.report_error(1001, [course.name, str(course.semester), str(student.semester_no)]))
            self.custom_log(False,
                            "Checked semester of students and courses that requested by students and they are declined ")
            return False
        if not self.check_credit(self, student, course):
            student.add_error(self._error.report_error(1002, [course.name, str(student.transcript.completed_credit)]))
            self.custom_log(False, "Compared credit of students and required credit of course and it is not enough.")
            return False
        if not self.check_quota(self, course):
            student.add_error(self._error.report_error(1003, [course.name]))
            self.custom_log(False, "Check quota of course and there is not enough space for student.")
            return False
        temp_return = self.check_pre_requisite(self, student, course)
        if not temp_return[0]:
            student.add_error(self._error.report_error(1004, [course.name, temp_return[1]]))
            return False
        temp_return_2 = self.check_collision(self, student)
        if not temp_return_2[0]:
            student.add_error(self._error.report_error(1005, [temp_return_2[1], temp_return_2[2], temp_return_2[3]]))
            return False
        if not self.check_elective(self, student, course):
            return False
        if not self.fte_takeable(self, student, course):
            # return False
            pass
        self.custom_log(True, "courseAvailability method is called and advisor check the course availability.")

        return True

    def check_quota(self, course:Course):
        if course.get_registered_students_count() >= course.get_quota():
            return False
        return True

    def check_semester(self, student:Student, course:Course):
        if course.get_semester() <= student.semester_no:
            return True
        return False

    def check_credit(self, student:Student, course:Course):
        if course.get_required_credits() <= student.transcript.get_completed_credit():
            return True
        return False

    def check_pre_requisite(self, student:Student, course:Course):
        temp_completed_course = student.transcript.get_completed_courses()
        availability = True
        error_info = [None, None]
        for i in range(len(course.get_prerequisite_courses())):
            for j in range(len(temp_completed_course)):
                if course.get_prerequisite_courses()[i].get_code() != temp_completed_course[j].code:
                    availability = False
                    error_info[1] = course.get_prerequisite_courses()[i].get_code()
                else:
                    availability = True
                    error_info[1] = ""
                    break
            if not availability:
                break
        error_info[0] = availability
        return error_info

    def check_collision(self, student:Student):
        courses_of_hash = student.selected_courses.keys()
        courses = list(courses_of_hash)
        error_info = [None, None, None, None]
        error_info[0] = True
        error_info[1] = ""
        error_info[2] = ""
        error_info[3] = ""
        for i in range(len(courses)):
            temp = list(courses[i].weekly_hours)
            for j in range(i + 1, len(courses)):
                temp = [x for x in temp if x in courses[j].weekly_hours]
                if len(temp) != 0:
                    error_info[0] = False
                    error_info[1] = courses[i].name
                    error_info[2] = courses[j].name
                    error_info[3] = str(len(temp))
                    return error_info
        return error_info

    # check elective calışmıyor
    def check_elective(self, student:Student, course:Course):
        count = 0
        for k, v in student.selected_courses.items():
            if isinstance(k, Elective):
                count += 1
        if count >= 2:
            return False
        return True

    # fte takeable calışmıyor

    def fte_takeable(self, student:Student, course:Course):
        if student.semester_no >= 7 and isinstance(course, Elective) and course.type == ElectiveType.FTE:
            return True
        return False

    @property
    def students(self):
        return self.__students

    @students.setter
    def students(self, students):
        self.__students = students

    def to_json(self):
        return {}
